### PR TITLE
12 hour support, AM/PM support

### PR DESCRIPTION
**Pull request by benjaminflopez21**
I added 12 hour support for native,

you just have to add the twelvehour="true" attribute to mat-datetimepicker tag
example: <mat-datetimepicker #timeAMPMPicker twelvehour="true" type="time">

### CODE REVIEW

File: projects/core/src/datetimepicker/calendar.scss [hunk 1]
```diff
@@ -1,173 +1,162 @@
...
-
 .mat-datetimepicker-calendar {
-  @include user-select(none);
-  display: block;
-  outline: none;
-
-  &[mode='landscape'] {
-    display: flex;
-  }
+    @include user-select(none);
```
Comment (bbortt) [inline projects/core/src/datetimepicker/calendar.scss:11]: please stick to two space indentation. I cannot see your changes this way!

File: projects/core/src/datetimepicker/calendar.html [hunk 2]
```diff
@@ -1,72 +1,41 @@
...
-  </mat-datetimepicker-clock>
-</div>
+    <mat-datetimepicker-month-view *ngSwitchCase="'month'" [activeDate]="_activeDate" [type]="type" [selected]="selected" [dateFilter]="_dateFilterForViews" (selectedChange)="_dateSelected($event)" (_userSelection)="_userSelected()">
+    </mat-datetimepicker-month-view>
+    <mat-datetimepicker-year-view *ngSwitchCase="'year'" [activeDate]="_activeDate" [type]="type" [selected]="selected" [dateFilter]="_dateFilterForViews" (selectedChange)="_monthSelected($event)" (_userSelection)="_userSelected()">
+    </mat-datetimepicker-year-view>
+    <mat-datetimepicker-clock *ngSwitchDefault [twelvehour]="twelvehour" [startView]="_clockView" [interval]="timeInterval" [minDate]="minDate" [maxDate]="maxDate" [dateFilter]="dateFilter" [selected]="_activeDate" (activeDateChange)="_onActiveDateChange($event)"
+        (selectedChange)="_timeSelected($event)" (_userSelection)="_userSelected()">
+    </mat-datetimepicker-clock>
+</div>
```
Comment (bbortt) [inline projects/core/src/datetimepicker/calendar.html:41]: it's hard to see the actual changes if you reformat the whole file.. please do *not* do this!
line breaks were there for a reason. it's very unpleasant having to scroll to the right if you have that long lines.

File: projects/core/src/datetimepicker/calendar.ts [hunk 3]
```diff
@@ -194,7 +198,14 @@ export class MatDatetimepickerCalendar<D> implements AfterContentInit, OnDestroy
...
   get _hoursLabel(): string {
-    return this._2digit(this._adapter.getHour(this._activeDate));
+    let hour = this._adapter.getHour(this._activeDate);
+    if (this.twelvehour) {
+      if (hour === 0) {
+        hour = 24;
+      }
+      hour = hour > 12 ? (hour - 12) : hour;
+    }
+    return this._2digit(hour);
```
Comment (bbortt) [inline projects/core/src/datetimepicker/calendar.ts:208]: this part seems ok 👍

File: projects/core/src/datetimepicker/datetimepicker-content.html [hunk 4]
```diff
@@ -1,15 +1,4 @@
...
-              [selected]="datetimepicker._selected"
-              [startAt]="datetimepicker.startAt"
-              (selectedChange)="datetimepicker._select($event)"
-              (viewChanged)="datetimepicker._viewChanged($event)"
-              (_userSelection)="datetimepicker.close()">
-</mat-datetimepicker-calendar>
+<mat-datetimepicker-calendar class="mat-typography" cdkTrapFocus [id]="datetimepicker.id" [attr.mode]="datetimepicker.mode" [startView]="datetimepicker.startView" [type]="datetimepicker.type" [twelvehour]="datetimepicker.twelvehour" [timeInterval]="datetimepicker.timeInterval"
+    [minDate]="datetimepicker._minDate" [maxDate]="datetimepicker._maxDate" [dateFilter]="datetimepicker._dateFilter" [selected]="datetimepicker._selected" [startAt]="datetimepicker.startAt" (selectedChange)="datetimepicker._select($event)" (viewChanged)="datetimepicker._viewChanged($event)"
+    (_userSelection)="datetimepicker.close()">
+</mat-datetimepicker-calendar>
```
Comment (bbortt) [inline projects/core/src/datetimepicker/datetimepicker-content.html:4]: again, do *not* reformat the whole file please. formatting is important and (most of the time) was done for a reason.

File: projects/core/src/datetimepicker/datetimepicker.ts [hunk 5]
```diff
@@ -98,6 +98,7 @@ export class MatDatetimepickerContent<D> implements AfterContentInit {
   preserveWhitespaces: false
 })
 export class MatDatetimepicker<D> implements OnDestroy {
+  @Input() twelvehour: boolean = false;
```
Comment (bbortt) [inline projects/core/src/datetimepicker/datetimepicker.ts:101]: maybe add reasonable comment like on line `102`.

File: src/app/date.component.html [hunk 6]
```diff
@@ -1,56 +1,63 @@
...
+        <mat-error *ngIf="group.get('filtertest').errors?.required">required</mat-error>
+        <mat-error *ngIf="group.get('filtertest').errors?.matDatepickerFilter">filter</mat-error>
+    </mat-form-field>
+    <mat-form-field>
+        <mat-placeholder>TouchUi</mat-placeholder>
+        <input matInput formControlName="touch" [min]="min" [matDatetimepicker]="touch" required>
+        <mat-datetimepicker-toggle [for]="touch" matSuffix></mat-datetimepicker-toggle>
+        <mat-datetimepicker #touch type="datetime" openOnFocus="true" mode="portrait" timeInterval="5" touchUi="true"></mat-datetimepicker>
+    </mat-form-field>
+</form>
```
Comment (bbortt) [inline src/app/date.component.html:63]: two space indentation.

File: projects/core/src/datetimepicker/calendar.html (hunk 2 above)
Comment (benjaminflopez21) [reply projects/core/src/datetimepicker/calendar.html:82]: Done

File: projects/core/src/datetimepicker/calendar.scss (hunk 1 above)
Comment (benjaminflopez21) [reply projects/core/src/datetimepicker/calendar.scss]: Done

File: projects/core/src/datetimepicker/calendar.html (hunk 2 above)
Comment (benjaminflopez21) [reply projects/core/src/datetimepicker/calendar.html:82]: Done

File: projects/core/src/datetimepicker/calendar.ts (hunk 3 above)
Comment (benjaminflopez21) [reply projects/core/src/datetimepicker/calendar.ts:208]: Cool :)

File: projects/core/src/datetimepicker/datetimepicker-content.html (hunk 4 above)
Comment (benjaminflopez21) [reply projects/core/src/datetimepicker/datetimepicker-content.html]: Done

File: projects/core/src/datetimepicker/datetimepicker.ts (hunk 5 above)
Comment (benjaminflopez21) [reply projects/core/src/datetimepicker/datetimepicker.ts:102]: Comment added.

File: src/app/date.component.html (hunk 6 above)
Comment (benjaminflopez21) [reply src/app/date.component.html]: Done

File: projects/core/src/datetimepicker/calendar.scss [hunk 7]
```diff
@@ -1,20 +1,16 @@
 @import '~@angular/material/theming';
-
```
Comment (bbortt) [inline projects/core/src/datetimepicker/calendar.scss:2]: just a hint for future contributions.. to any project..
such changes (reformattings, removal of line breaks, etc..) are often not very liked to see.
* formatting was done for a reason
* it makes reviewing PRs hard
* it's kind of a "i dont like your style guides" (which is.. well, not quit friendly)
I can overlook at this time.. but, if you find the time: please undo it ;)

File: projects/core/src/datetimepicker/calendar.scss [hunk 8]
```diff
@@ -42,17 +37,21 @@ $mat-calendar-next-icon-transform: translateX(-2px) rotate(45deg);
...
   line-height: 34px;
-
   [mode='landscape'] & {
     white-space: normal;
     word-wrap: break-word;
   }
 }
 
+.mat-datetimepicker-calendar-header-ampm-container {
+  font-size: 23px;
```
Comment (bbortt) [inline projects/core/src/datetimepicker/calendar.scss:47]: hard-coded size in pixels is not adaptable to screen size. what was the reason for this? could you maybe use `em`?

File: projects/core/src/datetimepicker/calendar.html [hunk 9]
```diff
@@ -17,6 +17,15 @@
...
                                                                      (click)="_minutesClicked()">{{ _minutesLabel }}</span>
+      <br/>
+      <span *ngIf="twelvehour" 
+            class="mat-datetimepicker-calendar-header-ampm-container">
+        <span class="mat-datetimepicker-calendar-header-ampm" 
+              [class.active]="_AMPM === 'AM'" 
+              (click)="_ampmClicked('AM')">AM</span>/<span class="mat-datetimepicker-calendar-header-ampm" 
+                                                            [class.active]="_AMPM === 'PM'" 
+                                                            (click)="_ampmClicked('PM')">PM</span>
+      </span>
```
Comment (bbortt) [inline projects/core/src/datetimepicker/calendar.html:28]: thanks for the example. this is very good!

File: projects/core/src/datetimepicker/calendar.ts [hunk 10]
```diff
@@ -194,7 +198,14 @@ export class MatDatetimepickerCalendar<D> implements AfterContentInit, OnDestroy
   }
 
   get _hoursLabel(): string {
-    return this._2digit(this._adapter.getHour(this._activeDate));
+    let hour = this._adapter.getHour(this._activeDate);
+    if (this.twelvehour) {
```
[the remark this answers is not in the document]
Comment (benjaminflopez21) [reply projects/core/src/datetimepicker/calendar.ts]: Done.

File: projects/core/src/datetimepicker/clock.ts [hunk 11]
```diff
@@ -122,7 +122,15 @@ export class MatDatetimepickerClock<D> implements AfterContentInit {
   _selectedMinute: number;
 
   get _hand(): any {
-    this._selectedHour = this._adapter.getHour(this.activeDate);
+    let hour = this._adapter.getHour(this.activeDate);
+    if (this.twelvehour) {
```
[the remark this answers is not in the document]
Comment (benjaminflopez21) [reply projects/core/src/datetimepicker/clock.ts]: Done.

File: projects/core/src/datetimepicker/calendar.scss (hunk 7 above)
Comment (benjaminflopez21) [reply projects/core/src/datetimepicker/calendar.scss]: Sorry about that, it was not my intention, my vscode was configured to format on save, but I reformat the file following your style guide.

File: projects/core/src/datetimepicker/calendar.scss (hunk 8 above)
Comment (benjaminflopez21) [reply projects/core/src/datetimepicker/calendar.scss]: Done

File: projects/core/src/datetimepicker/calendar.html (hunk 9 above)
Comment (benjaminflopez21) [reply projects/core/src/datetimepicker/calendar.html:28]: Cool, :+1:

**Pull request merged**